src/editor/autocomplete.js

```diff
@@ -0,0 +1,97 @@
+/*
+Copyright 2019 New Vector Ltd
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import {UserPillPart, RoomPillPart, PlainPart} from "./parts";
+
+export default class AutocompleteWrapperModel {
+    constructor(updateCallback, getAutocompleterComponent, updateQuery) {
+        this._updateCallback = updateCallback;
+        this._getAutocompleterComponent = getAutocompleterComponent;
+        this._updateQuery = updateQuery;
+        this._query = null;
+    }
+
+    onEscape(e) {
+        this._getAutocompleterComponent().onEscape(e);
+        this._updateCallback({
+            replacePart: new PlainPart(this._queryPart.text),
+            caretOffset: this._queryOffset,
+            close: true,
+        });
+    }
+
+    onEnter() {
+        this._updateCallback({close: true});
+    }
+
+    onTab() {
+        //forceCompletion here?
```

Comment:
unresolved TODO

Reply:
Indeed. This is not fully featured yet, but I want to get this in front of people testing it sooner rather than later.